Dual frames part1: Enable serialization of functions of time

nilsdeppe opened the pull request:
## Proposed changes

Functions of time need to be serializable for dual frames (because everything needs to be serializable...). This PR makes the already implemented serializable.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Enable serialization of functions of time

fmahebert commented:
For my understanding: what is the relationship between serialization and dual frames? I understand that we need everything to be serializable for checkpoint/restarts and for load balancing, where data may need to be moved around between nodes on the machine. What is it about dual frames that requires serialization?

nilsdeppe commented:
In order to get the time-dependent maps and functions of time into the DataBox they need to be serializable. So it's not needed from a moving mesh perspective but from an implementation perspective. Does that clarify anything?

fmahebert commented:
Yes, thank you 🙂

kidder commented:
@fmahebert since you've looked at this at least briefly, could you provide a review?

fmahebert commented on tests/Unit/ControlSystem/Test_PiecewisePolynomial.cpp:
```diff
@@ -1,101 +1,79 @@
 // Distributed under the MIT License.
 // See LICENSE.txt for details.
 
-#include "tests/Unit/TestingFramework.hpp"
```
Per our guidelines, this include has to come first. Why is it being moved?

nilsdeppe replied:
Ugh... I think this is a bug in clang format 9...

nilsdeppe replied:
Oh, it's a feature where they changed the default behavior or something... I've pushed a new commit that maintains the behavior we've had so far.

fmahebert commented on tests/Unit/ControlSystem/Test_SettleToConstant.cpp:
```diff
@@ -1,32 +1,20 @@
 // Distributed under the MIT License.
 // See LICENSE.txt for details.
 
-#include "tests/Unit/TestingFramework.hpp"
```
Again

fmahebert commented on tests/Unit/ControlSystem/Test_PiecewisePolynomial.cpp:
```diff
@@ -112,6 +90,112 @@ SPECTRE_TEST_CASE(
+          serialize_and_deserialize(f_of_t);
+
+      test(make_not_null(&f_of_t), make_not_null(&f_of_t), t, dt, final_time);
+      test(make_not_null(&f_of_t2), make_not_null(&f_of_t2), t, dt, final_time);
```
I would find this test (and the others) much cleaner if instead of checking that the two `f_of_t`'s behave the same, you checked that they were actually identical using an `operator==`. Isn't this how we test serialization in most of the code? I find that separating the test of the behavior from the test of the serialization helps modularity and understandability of the test.

Is there a particular reason why a `FunctionOfTime` shouldn't have an `operator==`?

nilsdeppe replied:
The derived classes could have `operator==`, though they're not copyable so it doesn't really make sense to have them comparable. Additionally, we've had a number of times where `operator==` was implemented incorrectly, so the behavioral test is value in either case.

fmahebert replied:
Okay, sounds like reason enough 👍

fmahebert commented on tests/Unit/ControlSystem/Test_PiecewisePolynomial.cpp:
```diff
@@ -1,101 +1,79 @@
   while (t < final_time) {
     t += dt;
-    f_of_t_derived.update(t, {3.0 + t});
+    f_of_t_derived->update(t, {3.0 + t});
```
I know this is neither your code or new code, so this shouldn't hold up the PR... but I don't understand what is happening here. Why is there a separate pointer to the `f_of_t`, of the derived type, for calling `update`?

nilsdeppe replied:
I don't remember the conversation exactly, but my guess is that it tests that the updating via the base class works correctly.

nilsdeppe committed:
Format include blocks separately

With clang-format 9 we need to explicitly say we don't want include blocks
merged during formatting.
